autopush-common/src/db/bigtable/bigtable_client/mod.rs

```diff
@@ -0,0 +1,1225 @@
+                    "Invalid DSN: Unparsable host {:?}",
+                    endpoint
+                )))?,
+            parsed.port().unwrap_or(8086)
+        );
+        if !parsed.path().is_empty() {
+            return Err(DbError::ConnectionError(format!(
+                "Invalid DSN: Table paths belong in settings : {:?}",
+                endpoint
+            )));
+        }
+        let db_settings = BigTableDbSettings::try_from(settings.db_settings.as_ref())?;
+        debug!("🉑 {:#?}", db_settings);
+        let mut chan = ChannelBuilder::new(env)
+            .max_send_message_len(1 << 28)
+            .max_receive_message_len(1 << 28);
+        // Don't get the credentials if we are running in the emulator
+        if settings
+            .dsn
+            .clone()
+            .map(|v| v.contains("localhost"))
+            .unwrap_or(false)
+            || std::env::var("BIGTABLE_EMULATOR_HOST").is_ok()
+        {
+            debug!("🉑 Using emulator");
+        } else {
+            chan = chan.set_credentials(
+                ChannelCredentials::google_default_credentials()
+                    .map_err(|e| DbError::ConnectionError(e.to_string()))?,
+            );
+            debug!("🉑 Using real");
+        }
+
+        let con_str = format!("{}{}", origin, parsed.path());
+        debug!("🉑 connection string {}", &con_str);
+        let chan = chan.connect(&con_str);
+        let client = BigtableClient::new(chan.clone());
+        Ok(Self {
+            settings: db_settings,
+            client,
+            chan,
+            _metrics: metrics,
+        })
+    }
+
+    /// Read a given row from the row key.
+    async fn read_row(
+        &self,
+        row_key: &str,
+        timestamp_filter: Option<u64>,
+    ) -> Result<Option<row::Row>, error::BigTableError> {
+        debug!("🉑 Row key: {}", row_key);
+
+        let mut row_keys = RepeatedField::default();
+        row_keys.push(row_key.to_owned().as_bytes().to_vec());
+
+        let mut row_set = data::RowSet::default();
+        row_set.set_row_keys(row_keys);
+
+        let mut req = bigtable::ReadRowsRequest::default();
+        req.set_table_name(self.settings.table_name.clone());
+        req.set_rows(row_set);
+
+        let rows = self.read_rows(req, timestamp_filter, None).await?;
+        Ok(rows.get(row_key).cloned())
+    }
+
+    /// Take a big table ReadRowsRequest (containing the keys and filters) and return a set of row data indexed by row key.
+    ///
+    ///
+    async fn read_rows(
+        &self,
+        req: ReadRowsRequest,
+        timestamp_filter: Option<u64>,
+        limit: Option<usize>,
+    ) -> Result<BTreeMap<RowKey, row::Row>, error::BigTableError> {
+        let resp = self
+            .client
+            .clone()
+            .read_rows(&req)
+            .map_err(|e| error::BigTableError::Read(e.to_string()))?;
+        merge::RowMerger::process_chunks(resp, timestamp_filter, limit).await
```

Comment:
Yes, but remember that the `row_limit` function limits rows regardless of what filtering we are doing. If we fetch the first 10 of 20 candidate rows, and 5 are excluded due to our limits, we will only return the remaining 5 and no additional records. This can be a problem since, while the `sortkey_timestamp` is part of the primary key and _should_ be used as the internal sorting key used by Bigtable, we've previously noted that both `range_key` and `regex` keys are potentially problematic. 
(in the case of `range_key`, it returns *all* rows contained between the start and stop ranges, requiring very explicit key construction, and `regex_key` does not provide syntax for "starting with keys matching the following pattern".) 

All that said, I've enabled row limits for now and removed the `limit` argument pass. Tests are currently passing, but I'm still feeling cautious about this.

Reply:
Just to clarify -- this is only an issue for the filtering we do on *our* side, right? (just the `timestamp_filter` against `sortkey_timestamp`)

Reply:
Correct. This is just for filtering in autopush code, not in bigtable. Bigtable filtering may still be possible, but would require additional investigation (and I'm not sure it would do what we want).

Reply:
Sorry, I think we'll need to leave it (at least for now) then -- we don't have a test that produces enough notifications to trigger the case of it not collecting enough notifications. I was able to create one by duplicating the integration test's `test_ttl_batch_partly_expired_and_good_one` and changing its `range(6)` calls to `range(12)`. Fails on bigtable but passes on dynamo.

Reply:
If we're able to switch to read_rows w/ a start_key instead of the regex matching (later) then I think we can begin taking advantage of the server side limit here.

Reply:
Totally fair. That was pretty much the reason I had done the initial version that way. I just thought you might remember some exclusion that would allow us to use that fix.